doc/specs/Keyboard-Selection.md

```diff
@@ -0,0 +1,173 @@
+It may seem that some effort can be saved by making the keyboard selection act as a simulation of mouse selection. There is a union of mouse and keyboard activity that can be represented in a single set of selection motion interfaces that are commanded by the TermControl's Mouse/Keyboard handler and adapted into appropriate motions in the Terminal Core.
+
+However, the mouse handler operates by translating a pixel coordinate on the screen to a text buffer coordinate. This would have to be rewritten and the approach was deemed unworthy.
+
+### Fundamental Terminal Core Changes
+
+The Terminal Core will need to expose a `KeyboardSelection()` function that is called by the keybinding handler. The following parameters will need to be passed in:
+- `enum Direction`: the direction that the selection anchor will attempt to move to. Possible values include `Up`, `Down`, `Left`, and `Right`.
+- `enum SelectionExpansionMode`: the selection expansion mode that the selection anchor will adhere to. Possible values include `Cell`, `Word`, and `Viewport`.
+
+For `SelectionExpansionMode = Cell`, the selection anchor will be updated according to the buffer's output pattern. For **horizontal movements**, the selection anchor will attempt to move left or right. If a viewport boundary is hit, the anchor will wrap appropriately (i.e.: hitting the left boundary moves it to the last cell of the line above it).
+
+For **vertical movements**, the selection anchor will attempt to move up or down. If a **viewport boundary** is hit and there is a scroll buffer, the anchor will move and scroll accordingly by a line. If a **buffer boundary** is hit, the anchor will not move. In this case, however, the event will still be considered handled.
+
+For `SelectionExpansionMode = Word`, the selection anchor will also be updated according to the buffer's output pattern, as above. However, the selection will be updated in accordance with "chunk selection" (performing a double-click and dragging the mouse to expand the selection). For **horizontal movements**, the selection anchor will be updated according to the `_ExpandDoubleClickSelection` functions. The result must be saved to the anchor. As before, if a boundary is hit, the anchor will wrap appropriately.
+
+For **vertical movements**, the movement is a little more complicated than before. The selection will still respond to buffer and viewport boundaries as before. If the user is trying to move up, the selection anchor will attempt to move up by one line, then selection will be expanded leftwards. Alternatively, if the user is trying to move down, the selection anchor will attempt to move down by one line, then the selection will be expanded rightwards.
+
+For `SelectionExpansionMode = Viewport`, the selection anchor will be updated according to the viewport's height. Horizontal movements will be updated according to the viewport's width, thus resulting in the anchor being moved to the left/right boundary of the viewport.

```

Comment:
Just making sure I understand, does this mean if a user hits left arrow while having the cursor on the right side of the viewport, it'll select the whole line from right to left? Similarly, if a user hits up arrow while at the bottom of the viewport, it'll select pretty much everything visible?

Also, since it's selection by viewport, that means pressing any direction while hitting a viewport boundary won't scroll the viewport up/down/left/right?

Reply:
> Just making sure I understand, does this mean if a user hits left arrow while having the cursor on the right side of the viewport, it'll select the whole line from right to left?

Yup

> Similarly, if a user hits up arrow while at the bottom of the viewport, it'll select pretty much everything visible?
> Also, since it's selection by viewport, that means pressing any direction while hitting a viewport boundary won't scroll the viewport up/down/left/right?

Kinda? It will have to scroll but only by one line. Here's an example:

```
Selection A --> B on a 5x5 viewport
-----
|     |
|     |
|     |
|     |
|  BA |
-----

Move B up by viewport results in...
|  B  |
-----
|     |
|     |
|     |
|     |
|   A |
-----
```

It'd be the equivalent of a page up/down.